server/odc-service/src/main/java/com/oceanbase/odc/service/flow/FlowSchedules.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2023 OceanBase.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.oceanbase.odc.service.flow;
+
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.apache.commons.collections4.CollectionUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Component;
+import org.springframework.transaction.support.TransactionTemplate;
+
+import com.oceanbase.odc.core.shared.constant.FlowStatus;
+import com.oceanbase.odc.metadb.flow.FlowInstanceRepository;
+import com.oceanbase.odc.metadb.flow.ServiceTaskInstanceEntity;
+import com.oceanbase.odc.metadb.flow.ServiceTaskInstanceRepository;
+import com.oceanbase.odc.metadb.task.TaskEntity;
+import com.oceanbase.odc.metadb.task.TaskRepository;
+import com.oceanbase.odc.service.flow.model.FlowNodeStatus;
+import com.oceanbase.odc.service.flow.task.model.FlowTaskProperties;
+import com.oceanbase.odc.service.flow.task.model.RuntimeTaskConstants;
+import com.oceanbase.tools.dbbrowser.util.DBSchemaAccessorUtil;
+
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * {@link FlowSchedules}
+ *
+ * @author yh263208
+ * @date 2024-05-22 15:33
+ * @since ODC-release_4.3.0
+ */
+@Slf4j
+@Component
+public class FlowSchedules {
+
+    private static final Integer OB_MAX_IN_SIZE = 2000;
+    @Autowired
+    private TransactionTemplate transactionTemplate;
+    @Autowired
+    private FlowTaskProperties flowTaskProperties;
+    @Autowired
+    private TaskRepository taskRepository;
+    @Autowired
+    private FlowInstanceRepository flowInstanceRepository;
+    @Autowired
+    private ServiceTaskInstanceRepository serviceTaskInstanceRepository;
+
+    @Scheduled(fixedDelayString = "${odc.flow.task.heartbeat-timeout-check-interval-millis:15000}")
+    public void cancelHeartbeatTimeoutFlow() {
+        long timeoutSeconds = this.flowTaskProperties.getHeartbeatTimeoutSeconds();
+        if (timeoutSeconds <= 0) {
+            return;
+        }
+        long minTimeoutSeconds = RuntimeTaskConstants.DEFAULT_TASK_CHECK_INTERVAL_SECONDS * 3;
```

Comment:
What's the interval of updating the heartbeat? Should we keep the updating interval less than the heartbeat timeout?

Reply:
yes, you are right.
the updating interval is set to `RuntimeTaskConstants.DEFAULT_TASK_CHECK_INTERVAL_SECONDS` and we set the min heartbeat timeout is more than 3 times of it